### PR TITLE
Add contrail nagios checks

### PR DESCRIPTION
**Pull request by n-pochet**
* Add nrpe-external-master interface to all (except contrail-openstack)
contrail charms
* Add nagios_context and nagios_servicegroups config options to all
(except contrail-openstack) contrail charms

The following nagios checks were added:
* contrail-agent:
  * Add nagios checks as subordinate
  * Add contrail-status check
* contrail-analytics:
  * Add contrail-status check
  * Add http check for port 8081
* contrail-analyticsdb:
  * Add contrail-status check
* contrail-controller:
  * Add contrail-status check
  * Add http check for port 8082
  * Add http check for port 8143

This was successfully tested with the following bundle:
```yaml
series: bionic
variables:
  auth-mode:           &auth-mode           'rbac'
  cloud-admin-role:    &cloud-admin-role    admin
  openstack-origin:    &openstack-origin    'distro'
  docker-registry:     &docker-registry     hub.juniper.net/contrail
  docker-user:         &docker-user         include-file://docker-user.txt
  docker-password:     &docker-password     include-file://docker-password.txt
  image-tag:           &image-tag           5.0.2-0.360-queens
machines:
  '0':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '1':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '2':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '3':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '4':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '10':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '11':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
  '20':
    constraints: arch=amd64 tags=os-nodes
    series: bionic
services:
  cinder:
    charm: cs:cinder
    num_units: 1
    options:
      openstack-origin: *openstack-origin
      block-device: None
      glance-api-version: 2
      worker-multiplier: 0.25
      action-managed-upgrade: True
    to: 
    - lxd:0
  glance:
    charm: cs:glance
    options:
      openstack-origin: *openstack-origin
      worker-multiplier: 0.25
      action-managed-upgrade: True
    num_units: 1
    to: 
    - lxd:1
  keystone:
    charm: cs:keystone
    num_units: 1
    options:
      openstack-origin: *openstack-origin
      admin-password: admin
      admin-role: admin
      worker-multiplier: 0.25
      preferred-api-version: 3
      token-provider: 'fernet'
      action-managed-upgrade: True
    to:
    - lxd:0
  mysql:
    charm: cs:percona-cluster
    num_units: 1
    options:
      max-connections: 1000
      innodb-buffer-pool-size: 512M
    to:
    - lxd:1
  neutron-api:
    charm: cs:neutron-api
    num_units: 1
    options:
      openstack-origin: *openstack-origin
      worker-multiplier: 0.25
      manage-neutron-plugin-legacy-mode: false
      action-managed-upgrade: True
    to:
    - 0
  nova-cloud-controller:
    charm: cs:nova-cloud-controller
    num_units: 1
    options:
      openstack-origin: *openstack-origin
      console-access-protocol: spice
      network-manager: Neutron
      worker-multiplier: 0.25
      scheduler-default-filters: "AggregateInstanceExtraSpecsFilter,NumInstancesFilter,AggregateNumInstancesFilter,RetryFilter,AvailabilityZoneFilter,CoreFilter,RamFilter,ComputeFilter,ComputeCapabilitiesFilter,ImagePropertiesFilter,ServerGroupAntiAffinityFilter,ServerGroupAffinityFilter,PciPassthroughFilter"
      action-managed-upgrade: True
    to:
    - lxd:1
  nova-compute:
    charm: cs:nova-compute
    num_units: 2
    options:
      openstack-origin: *openstack-origin
      enable-live-migration: true
      enable-resize: true
      libvirt-image-backend: qcow2
      migration-auth-type: ssh
      virt-type: kvm
      aa-profile-mode: disable
      action-managed-upgrade: True
    to:
    - '10'
    - '11'
  ntp:
    charm: cs:ntp-28
    num_units: 0
  openstack-dashboard:
    charm: cs:openstack-dashboard
    options:
      openstack-origin: *openstack-origin
      webroot: "/"
      neutron-network-l3ha: True
      neutron-network-lb: True
      neutron-network-firewall: False
      action-managed-upgrade: True
    num_units: 1
    to:
    - lxd:1
  rabbitmq-server:
    charm: cs:rabbitmq-server
    num_units: 1
    options:
      source: *openstack-origin
    to:
    - lxd:0
  # Contrail components
  contrail-openstack:
    charm: ./contrail-charms/contrail-openstack
    series: bionic
    num_units: 0
    options:
      docker-registry: *docker-registry
      docker-user: *docker-user
      docker-password: *docker-password
      image-tag: *image-tag
  contrail-agent:
    charm: ./contrail-charms/contrail-agent
    num_units: 0
    series: bionic
    options:
      docker-registry: *docker-registry
      docker-user: *docker-user
      docker-password: *docker-password
      image-tag: *image-tag
      vhost-gateway: auto
  contrail-analyticsdb:
    charm: ./contrail-charms/contrail-analyticsdb
    num_units: 3
    series: bionic
    constraints: mem=4096 cores=2 root-disk=50G
    to:
      - kvm:2
      - kvm:3
      - kvm:4
    options:
      cassandra-minimum-diskgb: "4"
      cassandra-jvm-extra-opts: "-Xms1g -Xmx2g"
      docker-registry: *docker-registry
      docker-user: *docker-user
      docker-password: *docker-password
      image-tag: *image-tag
  contrail-analytics:
    charm: ./contrail-charms/contrail-analytics
    num_units: 3
    series: bionic
    constraints: mem=4096 cores=2 root-disk=50G
    options:
      docker-registry: *docker-registry
      docker-user: *docker-user
      docker-password: *docker-password
      image-tag: *image-tag
    to:
      - kvm:2
      - kvm:3
      - kvm:4
  contrail-controller:
    charm: ./contrail-charms/contrail-controller
    series: bionic
    num_units: 3
    options:
      auth-mode: *auth-mode
      cloud-admin-role: *cloud-admin-role
      cassandra-minimum-diskgb: "4"
      cassandra-jvm-extra-opts: "-Xms1g -Xmx2g"
      docker-registry: *docker-registry
      docker-user: *docker-user
      docker-password: *docker-password
      image-tag: *image-tag
    to:
      - 2
      - 3
      - 4
  contrail-keystone-auth:
    charm: ./contrail-charms/contrail-keystone-auth
    series: bionic
    num_units: 1
    to:
      - lxd:0
  # Nagios
  nagios:
    charm: cs:nagios
    num_units: 1
    to:
      - 20
  nrpe-container:
    charm: cs:nrpe
    options:
      nagios_hostname_type: unit
      disk_root: ''
      load: ''
      swap: ''
      swap_activity: ''
      mem: ''
  nrpe-host:
    charm: cs:nrpe
    options:
      nagios_hostname_type: host
      xfs_errors: "30"
relations:
  # openstack
  - [ "nova-compute:amqp", "rabbitmq-server:amqp" ]
  - [ "keystone:shared-db", "mysql:shared-db" ]
  - [ "nova-cloud-controller:identity-service", "keystone:identity-service" ]
  - [ "glance:identity-service", "keystone:identity-service" ]
  - [ "neutron-api:identity-service", "keystone:identity-service" ]
  - [ "neutron-api:shared-db", "mysql:shared-db" ]
  - [ "neutron-api:amqp", "rabbitmq-server:amqp" ]
  - [ "glance:shared-db", "mysql:shared-db" ]
  - [ "glance:amqp", "rabbitmq-server:amqp" ]
  - [ "nova-cloud-controller:image-service", "glance:image-service" ]
  - [ "nova-compute:image-service", "glance:image-service" ]
  - [ "nova-cloud-controller:cloud-compute", "nova-compute:cloud-compute" ]
  - [ "nova-cloud-controller:amqp", "rabbitmq-server:amqp" ]
  - [ "openstack-dashboard:identity-service", "keystone:identity-service" ]
  - [ "nova-cloud-controller:shared-db", "mysql:shared-db" ]
  - [ "nova-cloud-controller:neutron-api", "neutron-api:neutron-api" ]
  - [ "cinder:image-service", "glance:image-service" ]
  - [ "cinder:amqp", "rabbitmq-server:amqp" ]
  - [ "cinder:identity-service", "keystone:identity-service" ]
  - [ "cinder:cinder-volume-service", "nova-cloud-controller:cinder-volume-service" ]
  - [ "cinder:shared-db", "mysql:shared-db" ]
  - [ "ntp:juju-info", "nova-compute:juju-info" ]

  # contrail
  - [ "contrail-keystone-auth", "keystone" ]
  - [ "contrail-analytics", "contrail-analyticsdb" ]
  - [ "contrail-controller", "contrail-keystone-auth" ]
  - [ "contrail-controller", "contrail-analytics" ]
  - [ "contrail-controller", "contrail-analyticsdb" ]
  - [ "contrail-openstack", "nova-compute" ]
  - [ "contrail-openstack", "neutron-api" ]
  - [ "contrail-openstack", "contrail-controller" ]
  - [ "contrail-agent:juju-info", "nova-compute:juju-info" ]
  - [ "contrail-agent", "contrail-controller" ]
  - [ "contrail-controller", "ntp" ]
  - [ "contrail-analytics", "ntp" ]
  - [ "contrail-analyticsdb", "ntp" ]
  # Nagios
  - [ nrpe-container, contrail-keystone-auth ]
  - [ nrpe-host, contrail-controller]
  - [ nrpe-host, contrail-analytics]
  - [ nrpe-host, contrail-agent]
  - [ nrpe-host, contrail-analyticsdb]
  - [ nrpe-host, nova-compute]
  - [ nagios, nrpe-container ]
  - [ nagios, nrpe-host]
```

### CODE REVIEW

**Comment by Andrey-mp** [comment]
Hi,
Can you please tell me who wants this functionality, why and when?

**Comment by n-pochet** [comment]
Hi @Andrey-mp ,

I'm working for Canonical and we have a customer that requires that feature.
We need this feature to give visibility, into nagios, over the health of the Contrail elements.
We would need this feature as soon as possible.
Thanks for you help!

File: contrail-agent/hooks/contrail_agent_hooks.py [hunk 1]
```diff
@@ -166,6 +173,50 @@ def upgrade_charm():
...
+    nrpe_compat.add_check(
+        shortname='check_contrail_status_agent',
+        description='Check contrail-status',
+        check_cmd=check_contrail_status_cmd
+    )
+
+    nrpe_compat.write()
+
+
+def rsync_nrpe_checks(plugins_dir):
```
Comment (Andrey-mp) [inline contrail-agent/hooks/contrail_agent_hooks.py]: i see that this func and next are similar in all files - can you please move them into common_utils.py?

Comment (n-pochet) [reply contrail-agent/hooks/contrail_agent_hooks.py]: All common code is now is `common_utils.py`

**Comment by n-pochet** [comment]
@Andrey-mp I successfully retested a deployment with the updated charms from 69e818839c36f0d5be4ef2c3df315b3fc6ed39af

File: contrail-agent/hooks/common_utils.py [hunk 2]
```diff
@@ -201,3 +207,88 @@ def render_and_log(template, conf_file, ctx, perms=0o444):
...
+    nrpe_compat.add_check(
+        shortname=ctl_status_shortname,
+        description='Check contrail-status',
+        check_cmd=_contrail_status_cmd(name, plugins_dir)
+    )
+
+    nrpe_compat.write()
+
+
+def _rsync_nrpe_checks(plugins_dir):
```
Comment (Andrey-mp) [inline contrail-agent/hooks/common_utils.py]: I'm sorry - might be I was not so clear. But I asked just for two functions that are equal for all.
function update_nrpe_config has too much specific to be in common - please move it back.
function _contrail_status_cmd can be common only if will not contain check for 'analytics' and will pass cver to all scripts (script can use or ignore parameter)

**Comment by Andrey-mp** [comment]
all other logic looks good.
just one question - is it OK that script shows just first found error? would it be better to collect all issues and return all?

**Comment by n-pochet** [comment]
I reused the exact same logic as yours. It is more or less a copy-paste from https://github.com/Juniper/contrail-charms/blob/R5/contrail-agent/hooks/common_utils.py#L86-L124.
For a first implementation, it might be sufficient.
We might iterate on it and add more features afterwards.

**Comment by Andrey-mp** [comment]
> I reused the exact same logic as yours. It is more or less a copy-paste from https://github.com/Juniper/contrail-charms/blob/R5/contrail-agent/hooks/common_utils.py#L86-L124.
> For a first implementation, it might be sufficient.
> We might iterate on it and add more features afterwards.

OK.

**Comment by n-pochet** [comment]
The deployment is successful with the latest pushed commit.

File: contrail-agent/metadata.yaml [hunk 3]
```diff
@@ -9,6 +9,10 @@ series:
   - xenial
   - bionic
 subordinate: true
+provides:
+  nrpe-external-master:
+    interface: nrpe-external-master
+    scope: container
```
Comment (Andrey-mp) [inline contrail-agent/metadata.yaml]: is it OK to define scope container for subordinate charm?

Comment (n-pochet) [reply contrail-agent/metadata.yaml]: Yes. As described in the **Relations** section of the [Juju subordinate documentation page](https://docs.jujucharms.com/2.5/en/authors-subordinate-application)

**Pull request merged**